src/awkward/_do.py

```diff
@@ -247,7 +247,7 @@ def completely_flatten(
         return tuple(arrays)
 
 
-def flatten(layout: Content, axis: Integral = 1) -> Content:
+def flatten(layout: Content, axis: int = 1) -> Content:
```

Comment:
Why make this less general? A user might pass in an `axis` value derived by NumPy, such that its type might be `np.uint32` or something, rather than `int`. That should be expected of all `axis` types. I think you had a type alias for `axis` types, with or without accepting `None` as a possibility.

Reply:
Mypy doesn't support the `numbers.` type relations as type hints, so we have to be slightly less general. Right now, these hints are all internal, and we don't have any type hints (or many, if we do) at the L1 user-facing boundary in `ak.`, so we're only type checking internal-internal calls. 

My feelings here are that we'll sprinkle through some `cast` if we need to appease the type checker.